test-framework/common/src/main/java/io/quarkus/test/common/WithTestResource.java

```diff
@@ -0,0 +1,63 @@
+package io.quarkus.test.common;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Map;
+
+import io.quarkus.test.common.WithTestResource.List;
+
+/**
+ * Used to define a test resource.
+ *
+ * <b>All</b> {@code WithTestResource} annotations in the test module
+ * are discovered (regardless of the test which contains the annotation)
+ * and their corresponding {@code QuarkusTestResourceLifecycleManager}
+ * started <b>before</b> <b>any</b> test is run.
+ *
+ * Note that test resources are never restarted when running {@code @Nested} test classes.
+ * <p>
+ * This replaces {@link QuarkusTestResource}. The only difference is that the default value for
+ * {@link #restrictToAnnotatedClass()} {@code == true}.
```

Comment:
I think we need to make it the fact that one annotation applies to a single test by default more prominent.